src/sci/impl/reflector.cljc

```diff
@@ -0,0 +1,326 @@
+                      arg-type (if arg-types
+                                 (let [t (aget arg-types i)]
+                                   (if (and (nil? t) (some? arg))
+                                     (.getClass ^Object arg)
+                                     t))
+                                 (when arg (.getClass ^Object arg)))
+                      param-type (aget params i)]
+                  (if (param-arg-type-match? param-type arg-type)
+                    (recur (inc i))
+                    false))
+                true))))))
+
+#?(:clj
+   (defn- match-method
+     "Find the best matching method from a list of methods given args and optional arg-types."
+     [methods args arg-types]
+     (loop [methods methods
+            found-m nil]
+       (if-let [^Method m (first methods)]
+         (let [params (.getParameterTypes m)]
+           (if (and (is-congruent? params args arg-types)
+                    (or (nil? found-m)
+                        (Compiler/subsumes params (.getParameterTypes ^Method found-m))))
+             (recur (rest methods) m)
+             (recur (rest methods) found-m)))
+         found-m))))
+
+#?(:clj
+   (defn- widen-boxed-args
+     "Widen boxed numeric arguments (e.g., Integer -> Long, Float -> Double)."
+     [^objects args]
+     (let [widened (object-array (alength args))]
+       (dotimes [i (alength args)]
+         (let [arg (aget args i)]
+           (if (some? arg)
+             (let [val-class (.getClass ^Object arg)]
+               (aset widened i
+                     (cond
+                       (or (identical? val-class Integer) (identical? val-class Short) (identical? val-class Byte))
+                       (.longValue ^Number arg)
+
+                       (identical? val-class Float)
+                       (.doubleValue ^Number arg)
+
+                       :else
+                       arg)))
+             (aset widened i nil))))
+       widened)))
+
+#?(:clj
+   (defn invoke-matching-method
+     "Invoke a method matching the given name from a list of methods.
+      This is the core SCI-specific method that supports type hints via arg-types.
+      
+      Parameters:
+      - method-name: String name of the method
+      - methods: java.util.List of Method objects
+      - context-class: Class for error messages (can be nil for static methods)
+      - target: Object to invoke on (nil for static methods)
+      - args: Object array of arguments
+      - arg-types: Optional array of Class objects for type hints"
+     ([method-name methods ^Object target args]
+      (invoke-matching-method method-name methods
+                              (when target (.getClass target))
+                              target args nil))
+     ([method-name methods context-class target args]
+      (invoke-matching-method method-name methods context-class target args nil))
+     ([method-name methods context-class target args arg-types]
+      (let [methods (seq methods)]
```

Comment:
Should remove this shadowed binding `[methods (seq methods)]`, it makes us lose the `List` type for no reason.

Reply:
I fixed the code to not make a seq out of the (Array)List at all which is more performant. Thanks.